### PR TITLE
Raise error when affecting tuple moved into another segment.

### PR DESCRIPTION
**Pull request by kainwen**
6X already fixes the issue in the commit 64b8664. Since master
branch now has merged Postgres v12, we have all the basic tools
to fix this.

Split update (distribution key update) in Greenplum is implemented by
deleting a tuple from original segment and inserting it into new
segment.  This is similar to an update to partition key column in
upstream PostgreSQL, which is handled by commit f16241b.

Merging Postgres v12 (commit 19cd1cf4) already makes ExecDelete handle
split-update as upstream's UPDATE on partition keys. This commit just
changes the error message a little. With the new AM_API from upstream,
select-for-update is automatically solved, and a test case is added by
this commit.

Also, this commit removes the previous error-out logic in ExecDelete,
because:
  * if the second split-update is  a real UPDATE on distkeys, it will
    contain motion, then the following and thus will throw error when
    EvalPlanQual
  * if the second split-update is ORCA-generated UPDATE on non distkeys,
    it will invoke `table_lock_tuple`, and we have check there

Fixes GitHub issue #8919.

Discussion: https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/kET4SJ1AHBA/gauiHO2WAwAJ

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
@asimrp @zhangh43 @hlinnaka  Please help review this.

6X version commit is: https://github.com/greenplum-db/gpdb/commit/64b8664b193f136e3437088d7547927358efac67

Thanks a lot!

**Comment by kainwen** [comment]
travis-ci fails due to network issue, it cannot install packges. concourse pass and my dev full pipeline is green.

So let's just ignore travis-ci here.

**Review by asimrp — APPROVED** [review]

File: src/test/isolation2/expected/gdd/concurrent_update.out [hunk 1]
```diff
@@ -171,7 +171,7 @@ UPDATE 1
 1: end;
 END
 2<:  <... completed>
-ERROR:  concurrent updates distribution keys on the same row is not allowed  (seg1 127.0.1.1:7003 pid=108408)
+ERROR:  EvalPlanQual can not handle subPlan with Motion node  (seg1 127.0.1.1:7003 pid=34629)
```
Comment (asimrp) [inline src/test/isolation2/expected/gdd/concurrent_update.out:174]: This error message is not conveying much to the user.  The previous message makes much more sense.

Comment (asimrp) [reply src/test/isolation2/expected/gdd/concurrent_update.out:174]: Well, it's not simply an error message difference, I think, if we are using the upstream logic to denote a tuple moved to another partition to also denote a tuple moved to another segment, then the only diff we should have with upstream is:
```diff
--- a/src/backend/access/heap/heapam_handler.c                                                                                                                                                                                                 
+++ b/src/backend/access/heap/heapam_handler.c                                                                                                                                                                                                 
@@ -405,7 +405,7 @@ tuple_lock_retry:
                                if (ItemPointerIndicatesMovedPartitions(tid))                                                                                                                                                                  
                                        ereport(ERROR,                                                                                                                                                                                         
                                                        (errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),                                                                                                                                           
-                                                        errmsg("tuple to be locked was already moved to another partition due to concurrent update")));                                                                                       
+                                                        errmsg("tuple to be locked was already moved to another partition or segment due to concurrent update")));                                                                            
                                                                                                                                                                                                                                               
                                tuple->t_self = *tid;                                                                                                                                                                                          
                                if (heap_fetch(relation, &SnapshotDirty, tuple, &buffer))                                                                                                                                                      
```

**Review by asimrp — CHANGES_REQUESTED** [review]
Revoking my approval based on previous comment.

File: src/test/isolation2/expected/gdd/concurrent_update.out (hunk 1 above)
Comment (kainwen) [reply src/test/isolation2/expected/gdd/concurrent_update.out:174]: > This error message is not conveying much to the user. The previous message makes much more sense.

Hi @asimrp Thanks for your comments.

This message `EvalPlanQual can not handle subPlan with Motion node` is the direct result when I decide to remove the code here: https://github.com/greenplum-db/gpdb/pull/10919/files#diff-919a020fcc2f55c2914cde24111d6c52L951

We have several conditions that we need to abort second transaction, when we wake up and get the HeapUpdate result:
* the second transaction is split-update and contain motion, then evalplanqual will lead to this error message
* the second transaction is DELETE or UPDATE, it contains no motion, so when it try to lock the tuple again, it founds that it is split-update modified, so it abort.

So I remove previous code to abort in ExecDelete, we can just handle it in EvalPlanQual or heapam_tuple_lock.
After removing the code, the case  (sencond transaction is spilt-update with motion) will be handled in EvalPlanQual, that is the reason of the change of the error message.

**Comment by kainwen** [comment]
> Revoking my approval based on previous comment.

Hi, @asimrp 
Do my reply https://github.com/greenplum-db/gpdb/pull/10919#discussion_r498153358 make sense for you?

Thanks.

**Review by asimrp — APPROVED** [review]
Thank you for the explanation.  I get it now.  That error message needs to be dealt with separately.

**Pull request merged**